consensus/tendermint/adapter/store.go

```
@@ -82,11 +90,32 @@ func (s *Store) SaveBlock(block *types.FullBlock, commit *types.Commit) {
 
 // Validate a block without Commit and with LastCommit.
 func (s *Store) ValidateBlock(state pbft.ChainState, block *types.FullBlock) (err error) {
-	err = s.verifyHeaderFunc(s.chain, block.Header(), false)
+	header := block.Header()
+	err = s.verifyHeaderFunc(s.chain, header, false)
 	if err != nil {
 		return
 	}
 
+	remoteChainNumber := uint64(math.MaxUint64)
+	hash := common.Hash{}
+	l := len(prefix)
+	if len(header.Extra) >= l+8+32 && bytes.Compare(header.Extra[:l], prefix) == 0 {
```

I guess we should return the error directly if the remoteChainNumber is not found in header.Extra

---

if it is not an epoch header, the remoteChainNumber will be empty in header.Extra

---

If that is the case, we should process it explicitly - i.e., if not epoch header, then header.Extra must be empty; or if it is an epoch header, read and return err if not found.